engine/src/main/java/com/datatorrent/stram/client/AppPackage.java

```diff
@@ -255,19 +285,51 @@ private void processAppProperties()
    */
   public AppPackage(File file, boolean processAppDirectory) throws IOException, ZipException
   {
-    this(file, null, processAppDirectory);
+    this(new FileInputStream(file), processAppDirectory);
   }
 
-  public static void extractToDirectory(File directory, File appPackageFile) throws ZipException
+  /**
+   * Creates an App Package object.
+   *
+   * If app directory is to be processed, there may be resource leak in the class loader. Only pass true for short-lived
+   * applications
+   *
+   * Files in app package will be extracted to tmp folder and will be cleaned on close()
+   * The close() method could be explicitly called or implicitly called by GC finalize()
+   *
+   * @param input
+   * @param processAppDirectory
+   * @throws java.io.IOException
+   */
+  public AppPackage(InputStream input, boolean processAppDirectory) throws IOException
   {
-    ZipFile zipFile = new ZipFile(appPackageFile);
+    this(input, null, processAppDirectory);
+  }
 
-    if (zipFile.isEncrypted()) {
-      throw new ZipException("Encrypted app package not supported yet");
-    }
+  public static void extractToDirectory(File directory, File appPackageFile) throws IOException
+  {
+    extractToDirectory(directory, new JarInputStream(new FileInputStream(appPackageFile)));
+  }
 
-    directory.mkdirs();
-    zipFile.extractAll(directory.getAbsolutePath());
+  private static void extractToDirectory(File directory, JarInputStream input) throws IOException
+  {
+    File manifestFile = new File(directory, JarFile.MANIFEST_NAME);
+    manifestFile.getParentFile().mkdirs();
+    input.getManifest().write(new FileOutputStream(manifestFile));
+
+    JarEntry entry = input.getNextJarEntry();
+    while (entry != null) {
+      File newFile = new File(directory, entry.getName());
+      if (entry.isDirectory()) {
+        newFile.mkdirs();
+      } else {
+        FileOutputStream fileoutputstream = new FileOutputStream(newFile);
+        IOUtils.copy(input, fileoutputstream);
+        fileoutputstream.close();
```

Comment:
We can do this in try() - finally block to ensure outputstream is always closed.